metalearners/metalearner.py

```diff
@@ -119,15 +121,18 @@ def _combine_propensity_and_nuisance_specs(
 
 def _filter_x_columns(X: Matrix, feature_set: Features) -> Matrix:
     if feature_set is None:
-        X_filtered = X
-    elif len(feature_set) == 0:
-        X_filtered = np.ones((safe_len(X), 1))
-    else:
-        if isinstance(X, pd.DataFrame):
-            X_filtered = X[list(feature_set)]
-        else:
-            X_filtered = X[:, np.array(feature_set)]
-    return X_filtered
+        return X
+    if len(feature_set) == 0:
+        return np.ones((safe_len(X), 1))
+    if nw.dependencies.is_into_dataframe(X):
+        X_nw = nw.from_native(X)  # type: ignore
```

Comment:
`is_into_dataframe` will return True for eager frames, you can probably improve typing experience a bit by flagging the input as eager:

```suggestion
        X_nw = nw.from_native(X, eager_only=True)
```

Reply:
If I remove the imperative to tell mypy to ignore this and add `eager_only=True`, I get the following mypy complaint:

```
metalearners/metalearner.py:128: error: No overload variant of "from_native" matches argument types "Any | ndarray[Any, Any] | Any", "bool"  [call-overload]
metalearners/metalearner.py:128: note: Possible overload variants:
metalearners/metalearner.py:128: note:     def [IntoDataFrameT: NativeFrame | DataFrame[Any] | DataFrameLike] from_native(native_object: IntoDataFrameT | Series[Any] | NativeSeries, *, pass_through: Literal[True], eager_only: Literal[False] = ..., series_only: Literal[False] = ..., allow_series: Literal[True]) -> DataFrame[IntoDataFrameT]
metalearners/metalearner.py:128: note:     def [IntoDataFrameT: NativeFrame | DataFrame[Any] | DataFrameLike, IntoSeriesT: Series[Any] | NativeSeries] from_native(native_object: IntoDataFrameT | IntoSeriesT, *, pass_through: Literal[True], eager_only: Literal[True], series_only: Literal[False] = ..., allow_series: Literal[True]) -> DataFrame[IntoDataFrameT] | Series[IntoSeriesT]
metalearners/metalearner.py:128: note:     def [IntoDataFrameT: NativeFrame | DataFrame[Any] | DataFrameLike] from_native(native_object: IntoDataFrameT, *, pass_through: Literal[True], eager_only: Literal[False] = ..., series_only: Literal[False] = ..., allow_series: None = ...) -> DataFrame[IntoDataFrameT]
metalearners/metalearner.py:128: note:     def [T] from_native(native_object: T, *, pass_through: Literal[True], eager_only: Literal[False] = ..., series_only: Literal[False] = ..., allow_series: None = ...) -> T
metalearners/metalearner.py:128: note:     def [IntoDataFrameT: NativeFrame | DataFrame[Any] | DataFrameLike] from_native(native_object: IntoDataFrameT, *, pass_through: Literal[True], eager_only: Literal[True], series_only: Literal[False] = ..., allow_series: None = ...) -> DataFrame[IntoDataFrameT]
metalearners/metalearner.py:128: note:     def [T] from_native(native_object: T, *, pass_through: Literal[True], eager_only: Literal[True], series_only: Literal[False] = ..., allow_series: None = ...) -> T
metalearners/metalearner.py:128: note:     def [IntoFrameT: NativeFrame | DataFrame[Any] | LazyFrame[Any] | DataFrameLike, IntoSeriesT: Series[Any] | NativeSeries] from_native(native_object: IntoFrameT | IntoSeriesT, *, pass_through: Literal[True], eager_only: Literal[False] = ..., series_only: Literal[False] = ..., allow_series: Literal[True]) -> DataFrame[IntoFrameT] | LazyFrame[IntoFrameT] | Series[IntoSeriesT]
metalearners/metalearner.py:128: note:     def [IntoSeriesT: Series[Any] | NativeSeries] from_native(native_object: IntoSeriesT, *, pass_through: Literal[True], eager_only: Literal[False] = ..., series_only: Literal[True], allow_series: None = ...) -> Series[IntoSeriesT]
metalearners/metalearner.py:128: note:     def [IntoDataFrameT: NativeFrame | DataFrame[Any] | DataFrameLike] from_native(native_object: IntoDataFrameT, *, pass_through: Literal[False] = ..., eager_only: Literal[False] = ..., series_only: Literal[False] = ..., allow_series: None = ...) -> DataFrame[IntoDataFrameT]
metalearners/metalearner.py:128: note:     def [IntoDataFrameT: NativeFrame | DataFrame[Any] | DataFrameLike] from_native(native_object: IntoDataFrameT, *, pass_through: Literal[False] = ..., eager_only: Literal[True], series_only: Literal[False] = ..., allow_series: None = ...) -> DataFrame[IntoDataFrameT]
metalearners/metalearner.py:128: note:     def from_native(native_object: NativeFrame | DataFrame[Any] | LazyFrame[Any] | DataFrameLike | Series[Any] | NativeSeries, *, pass_through: Literal[False] = ..., eager_only: Literal[False] = ..., series_only: Literal[False] = ..., allow_series: Literal[True]) -> DataFrame[Any] | LazyFrame[Any] | Series[Any]
metalearners/metalearner.py:128: note:     def [IntoSeriesT: Series[Any] | NativeSeries] from_native(native_object: IntoSeriesT, *, pass_through: Literal[False] = ..., eager_only: Literal[False] = ..., series_only: Literal[True], allow_series: None = ...) -> Series[IntoSeriesT]
metalearners/metalearner.py:128: note:     def [IntoFrameT: NativeFrame | DataFrame[Any] | LazyFrame[Any] | DataFrameLike] from_native(native_object: IntoFrameT, *, pass_through: Literal[False] = ..., eager_only: Literal[False] = ..., series_only: Literal[False] = ..., allow_series: None = ...) -> DataFrame[IntoFrameT] | LazyFrame[IntoFrameT]
metalearners/metalearner.py:128: note:     def from_native(native_object: Any, *, pass_through: bool, eager_only: bool, series_only: bool, allow_series: bool | None) -> Any
Found 1 error in 1 file (checked 32 source files)
```

Reply:
I see, that's probably because we don't have type guards on `is_into_dataframe`. In any case, I would suggest to add `eager_only=True` even when keeping `# type: ignore`

Reply:
Great point. I added this in the calls to `from_native`. :)